tools/roc/application.e

```diff
@@ -0,0 +1,287 @@
+note
+	description: "[
+			roc tool: install modules into an existing CMS Application.
+			
+			roc install [--module|-m <MODULE_PATH>] [(--dir|-d <CMS_PATH>) | <MODULE_NAME>]
+
+			install:  Install a given module to the corresponding cms application
+			--module|-m: module path or current directory if is not defined.
+			--dir|-d cms application path or current directory if is not defined
+
+			Running the command will copy to the CMS Application site/modules the following artifacts if the current module provide them.
+
+				config
+				scripts
+				themes
+			running
+			roc instal blog
+			will look for a module blog in the modules directory starting at the current directory.
+	]"
+	date: "$Date$"
+	revision: "$Revision$"
+
+class
+	APPLICATION
+
+inherit
+
+	SHARED_EXECUTION_ENVIRONMENT
+
+	ARGUMENTS
+
+create
+	make
+
+feature {NONE} -- Initialization
+
+	make
+			-- Initialize tool.
+		do
+				-- TODO add support to other commands.
+			if argument_count = 0 then
+				print ("Use: roc install [--module|-m <MODULE_PATH>] [(--dir|-d <CMS_PATH>) | <MODULE_NAME>]")
+			elseif argument (1).starts_with ("install") then
+				execute_install
+			else
+				print ("Wrong command")
+				print ("%NUse: roc install [--module|-m <MODULE_PATH>] -cap <CMS_PATH>")
+			end
+		end
+
+feature {NONE} -- Install
+
+	execute_install
+			-- Install a new module.
+		local
+			i: INTEGER
+			error: BOOLEAN
+			optional_module: BOOLEAN
+			cms_path: BOOLEAN
+		do
+			if is_valid_install_cmd then
+				do_execute_install
+			end
+		end
+
+	do_execute_install
+			-- Install a module into a cms application.
+			-- Pattern
+			-- app/site/modules/module_name/config/....
+			-- app/site/modules/module_name/scripts/....
+			-- app/site/modules/module_name/themes/....
+		local
+			l_module_path: PATH
+			l_config: PATH
+			l_site_dir: DIRECTORY
+			l_modules_dir: DIRECTORY
+			l_dest_dir: DIRECTORY
+			l_file: FILE
+		do
+			if attached install_cms_path as l_cms_path then
+				l_module_path := install_module_path
+					--Install configuration files.
+				if attached l_module_path.entry as l_entry then
+					create l_site_dir.make_with_path (l_cms_path.extended ("site"))
+					create l_modules_dir.make_with_path (l_cms_path.extended ("site").extended ("modules"))
+
+					if
+						l_site_dir.exists and then
+					   l_modules_dir.exists
+					then
+					    create l_dest_dir.make_with_path (l_cms_path.extended ("site").extended ("modules").extended (l_entry.name))
+						if not l_dest_dir.exists then
+							l_dest_dir.create_dir
+						end
+					   	install_module_elements (l_module_path, l_dest_dir.path, Config_dir)
+						install_module_elements (l_module_path, l_dest_dir.path, Scripts_dir)
+						install_module_elements (l_module_path, l_dest_dir.path, Themes_dir)
+						print ("Module ")
+						print ( l_entry)
```

Comment:
if the l_entry is unicode ... output will be messed up.
you should use the encoding library, and the LOCALIZED_PRINTER ... annoying, but recommended for console app.